File: problems/树/110.平衡二叉树.py
#!/usr/bin/env python3
"""
110. 平衡二叉树
"""
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def is_balanced(root: Optional[TreeNode]) -> bool:
    def dfs(node: Optional[TreeNode]) -> int:
        if not node: return 0
        left = dfs(node.left)
        if left == -1: return -1
        right = dfs(node.right)
        if right == -1 or abs(left - right) > 1: return -1
        return max(left, right) + 1
    return dfs(root) != -1

if __name__ == "__main__":
    root1 = TreeNode(1, TreeNode(2, TreeNode(3), TreeNode(3)), TreeNode(2))
    print(is_balanced(root1))  # False
    root2 = TreeNode(1, TreeNode(2, TreeNode(3), TreeNode(3)), TreeNode(2, None, TreeNode(9)))
    print(is_balanced(root2))  # True
    print(is_balanced(None))  # True

File: problems/树/110.平衡二叉树.cpp
#include <iostream>
#include <algorithm>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

int height(TreeNode* node) {
    if (!node) return 0;
    int l = height(node->left);
    if (l == -1) return -1;
    int r = height(node->right);
    if (r == -1 || abs(l-r) > 1) return -1;
    return max(l, r) + 1;
}
bool isBalanced(TreeNode* root) { return height(root) != -1; }
int main() {
    TreeNode* root1 = new TreeNode(1);
    root1->left  = new TreeNode(2);
    root1->right = new TreeNode(2);
    root1->left->left  = new TreeNode(3);
    root1->left->right = new TreeNode(3);
    cout << (isBalanced(root1) ? "True" : "False") << endl;  // False
    TreeNode* root2 = new TreeNode(1);
    root2->left  = new TreeNode(2);
    root2->right = new TreeNode(2);
    root2->left->left  = new TreeNode(3);
    root2->left->right = new TreeNode(3);
    root2->right->right = new TreeNode(9);
    cout << (isBalanced(root2) ? "True" : "False") << endl;  // True
    cout << (isBalanced(nullptr) ? "True" : "False") << endl;  // True
    return 0;
}
